Improve Hide Content Options in Studio

efischer19 opened the pull request:
[TNL-4906](https://openedx.atlassian.net/browse/TNL-4906)

efischer19 commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -590,47 +590,65 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+                    publish: 'republish',
+                    metadata: {
+                        visible_to_staff_only: currentVisibility == 'staff_only' ? true : null,
+                        content_visibility: ['hide_after_due', 'hide_after_submit'].indexOf(currentVisibility) ? currentVisibility : null
```
@jcdyer FYI - I believe this is where I'll be passing xblock field data into the new backend.

efischer19 commented:
FYI @catong @sstack22 @chris-mike I've got the first pass of a studio front-end for hiding content functioning locally, here's a screenshot:

<img width="585" alt="screen shot 2016-07-06 at 4 04 08 pm" src="https://cloud.githubusercontent.com/assets/7373924/16632723/7360cd4a-4393-11e6-9e92-ad44e63b1d0f.png">

efischer19 commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -603,33 +617,63 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
+                // TODO: as part of TNL-4904, un-comment the below content_visibility lines
+                if (this.currentVisibility() === 'staff_only') {
+                    metadata.visible_to_staff_only = true;
+                    // metadata.content_visibility = null;
```
I believe this is where I'll be setting the `hide_after_due` field.

efischer19 commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -7,13 +7,14 @@
     'use strict';
     var CourseOutlineXBlockModal, SettingsXBlockModal, PublishXBlockModal, AbstractEditor, BaseDateEditor,
-        ReleaseDateEditor, DueDateEditor, GradingEditor, PublishEditor, StaffLockEditor,
+        ReleaseDateEditor, DueDateEditor, GradingEditor, PublishEditor, ContentVisibilityEditor,
```
Note to self: *subsections* are the only layer that needs a full `ContentVisibilityEditor`, sections and units should keep `StaffLockEditor` (or the same functionality, at least)

efischer19 commented on cms/templates/js/course-outline.underscore:
```diff
@@ -179,12 +179,10 @@ if (is_proctored_exam) {
                 <div class="status-hide-after-due">
                     <p>
-                        <% if (!is_proctored_exam && xblockInfo.get('hide_after_due')) { %>
+                        <!--TODO: make the following check use the new fields defined in TNL-4904-->
```
Unresolved TODO

efischer19 commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -1314,7 +1311,8 @@ define(["jquery", "edx-ui-toolkit/js/utils/spec-helpers/ajax-helpers", "common/j
                     outlinePage.$('.outline-item .outline-subsection .configure-button').click();
-                    setEditModalValues("7/9/2014", "7/10/2014", "Lab", true);
+                    setEditModalValues("7/9/2014", "7/10/2014", "Lab");
+                    //set staff_only here
```
Restore this

efischer19 commented:
There are very likely going to be js/bokchoy failures to be fixed after this round of tests.

efischer19 commented:
jenkins run python

efischer19 commented:
FYI @nasthagiri

chris-mike commented:
What is the default choice here?  Without a default choice, I wonder what happens if I select a radial and then decide to go back to the apparent default (i.e., nothing selected).

Also, I'm guessing @catong provided the language, so I assume it's consistent and believe it's clear. A small nit: I wonder about the plural form of "contents".  It seems the sentence could read with a singular "content" or "course content" to be more specific. Assuming "contents" is intentional, I have no problem with the plural.

efischer19 commented:
Sorry about that @chris-mike - I hadn't realized nothing was selected in that screenshot. It *should* be set up so that `Always show to students` is the default, and it should be impossible to de-select everything. It that a standard way of handling this situation?

I haven't gotten direct feedback from @catong on the language, so feedback appreciated there as well!

I agree that the default option needs to be called out more clearly - what would you recommend is the best way to do that? Maybe add `(Default)` at the end of that option?

efischer19 commented:
@sanfordstudent This is the studio companion to Nimisha's LMS PR linked above, would you be able to review it as well?

sanfordstudent commented:
@efischer19 taking a look.

efischer19 commented:
jenkins run python

sanfordstudent commented:
This looks really good. (+1)

efischer19 commented:
Something I hadn't considered until I just ran into it while setting up the sandbox - how should we handle the pre-existing `hide_after_due` setting for timed exams? It ought to be removed, yeah?

<img width="589" alt="screen shot 2016-07-15 at 4 00 09 pm" src="https://cloud.githubusercontent.com/assets/7373924/16886983/59bd09d0-4aa5-11e6-8eac-8d9c639f317d.png">

efischer19 commented:
@catong @sstack22 @chris-mike @sanfordstudent @nasthagiri I've got a sandbox set up with this working, the Studio link is https://studio-efischer19.sandbox.edx.org/course/course-v1:edX+DemoX+Demo_Course (bottom section).

Note that there's an unresolved issue with proctored exams and this functionality co-existing, we need to debug that a bit more on Monday. Otherwise, the changes outlined in https://openedx.atlassian.net/browse/TNL-4906 and the sub-task https://openedx.atlassian.net/browse/TNL-4951 are ready for feedback.

efischer19 commented:
@sanfordstudent - changes made since your thumbs are minor - I removed some functionality from the timed exam editor, as it is no longer setting `hide_after_due`, and fixed some css.

sanfordstudent commented:
👍

nasthagiri commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -605,43 +615,35 @@ define(["jquery", "edx-ui-toolkit/js/utils/spec-helpers/ajax-helpers", "common/j
                     this.$(".field-time-limit input").val(time_limit);
                     this.$(".field-time-limit input").trigger('focusout');
-                    this.$('.field-hide-after-due input').prop('checked', hide_after_due).trigger('change');
+                    if (hide_after_due) {
```
Should the content-visibility be updated when `hide_after_due` is `false`?

efischer19 replied:
I don't think so, but correct me if this line of thinking is wrong.

- Some timed exams already have `hide_after_due` set to True, but the default is false.
- With the changes being made in this PR, we will stop setting this field on timed exams (it won't be added to their database model). I need to double check to confirm this.
- `hide_after_due` is being set at the subsection level now

Given that, my thinking is that when a timed exam has `hide_after_due` set to True, it can override the xblock default visibility value. If it's not set, we should let the changes I'm making in this PR take precedence - having the default value of False is not an indicator that a timed exam is explicitly choosing to do anything with visibility.

efischer19 replied:
To follow up here:
- I've traced the code locally, and setting `hide_after_due` in the new fashion still sets the value in SQL for the timed exam, so we're all set there.
- For the test method specifically, the tests that are calling this are just verifying that selected options, when set, are returned from the modal on save. Thus, I've elected to always set `hide_after_due` visibility in `selectTimedExam`.

This comment will be lost in the diff after I push my latest commit, but I wanted to keep the conversation in context.

nasthagiri commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -915,38 +907,41 @@ define(["jquery", "edx-ui-toolkit/js/utils/spec-helpers/ajax-helpers", "common/j
                     
-                    // time limit and hide after due should be visible, review rules should be hidden
-                    checkOptionFieldVisibility(true, false, true);
+                    // time limit and should be visible, review rules should be hidden
```
remove extra "and"

nasthagiri commented on cms/templates/js/content-visibility-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+        <p class="tip tip-warning">
+            <% var message = gettext('If you make this %(xblockType)s visible to students, students will be able to see its content after the release date has passed and you have published the unit.'); %>
+            <%- interpolate(message, { xblockType: xblockType }, true) %>
+            <%- gettext('Any units that are explicitly hidden from students will remain hidden after you clear this option for the subsection.') %>
```
On this line, it explicitly says "subsection".  However, the line above parameterizes the xblock type.  Should be consistent with either approach.

catong replied:
Replace "students" with "learners" and +1 to Nimisha's comment.

nasthagiri commented on cms/templates/js/course-outline.underscore:
```diff
@@ -200,6 +189,14 @@ if (is_proctored_exam) {
+                <p>
+                    <% if (xblockInfo.get('hide_after_due')) { %>
+                        <span class="icon fa fa-eye-slash" aria-hidden="true"></span>
+                        <span class="status-hide-after-due-value"> <%- gettext("Subsection will remain hidden after due date") %> </span>
```
Nit: Should "Subsection" be parameterized here to the Assignment type of the subsection?  So it would say "Homework", "Lab", "Exam", etc?

efischer19 replied:
I'm not sure. @catong What do you think? The assignment type is more specific, but in my mind the user is already thinking in a section-subsection-unit level when this string is viewed.

![screen shot 2016-07-19 at 10 58 56 am](https://cloud.githubusercontent.com/assets/7373924/16954731/f3497604-4d9f-11e6-86d3-d6f2927db964.png)

nasthagiri replied:
FYI @sstack22

catong replied:
The "remain hidden" wording strikes me as a bit odd. I'd suggest striking the word "remain" and just saying "Subsection is hidden after due date".
@sstack22 @efischer19 @chris-mike @nasthagiri

nasthagiri commented on cms/templates/js/publish-xblock.underscore:
```diff
@@ -101,6 +101,10 @@ var visibleToStaffOnly = visibilityState === 'staff_only';
             </li>
         </ul>
+        <p>
+            <%- gettext("Note: Do not hide graded assignments using this feature.") %>
```
Is "this feature" clear to the user?  What feature, in particular, are we talking about here?

efischer19 replied:
It should be clear contextually, but I'm open to suggestions here.

Also @catong - I have a dummy doc link in place here, it'll need to be updated to point to real documentation before this merges. Any suggestions on where it should link?

![screen shot 2016-07-19 at 11 00 57 am](https://cloud.githubusercontent.com/assets/7373924/16954800/41dd2bb2-4da0-11e6-83b4-538484252c90.png)

nasthagiri replied:
Ok.  Gotcha.  Should we clarify the note to say not to use the feature for already released graded assignments?  For assignments that haven't been released yet, it's perfectly fine.  And we want to encourage them to use this feature for still in-development assignments, right?
`Do not hide graded assignments after they have been released.`

catong replied:
@efischer19 Based on this dev work, I'm going to rework existing sections in the course authors' guide, so I'll provide a link when I have the new structure, and will also suggest link text.

nasthagiri commented on common/test/acceptance/pages/studio/container.py:
```diff
@@ -191,10 +193,19 @@ def toggle_staff_lock(self, inherits_staff_lock=False):
         """
         was_locked_initially = self.is_staff_locked
+
+        # Sections and Units lock differently than subsections, so try both actions
```
Is it possible to be explicit about this by checking which modal is open (for subsection or not) and then clicking the exactly correct UI?

efischer19 replied:
I was wondering why we have to care about doing this check in 2 different files when I realized - this isn't the modal at all. It's the publish view of a single unit. My changes related to subsection content hiding don't apply here at all.

nasthagiri commented on common/test/acceptance/pages/studio/overview.py:
```diff
@@ -1037,19 +1027,48 @@ def policy(self, grading_label):
+        if needed.
+        """
+        if not self.is_staff_lock_visible:
+            self.find_css(".settings-tab-button[data-tab=advanced]").click()
```
I'm concerned this can cause flaky test behavior.  Here, we are not waiting for the advanced tab to be visible and ready before proceeding.  And since the code is written to work for either subsection and non-subsection blocks, it cannot be robust since it doesn't know which exact UI to expect.

How difficult would it be for the caller to pass down which UI it is expecting?  For subsection or for non-subsection?  If you do this, you can have smaller methods for each and eliminate the various `if` statements in the changes in this file.

efischer19 replied:
You're right - I'm not a fan of the subsection/nonsubsection logic in these tests. I'll clean that up.

nasthagiri commented on common/test/acceptance/pages/studio/overview.py:
```diff
@@ -1037,19 +1027,48 @@ def policy(self, grading_label):
+            return self.find_css('#staff_lock')[0].is_selected()
 
     @is_explicitly_locked.setter
     def is_explicitly_locked(self, value):
```
Should this method be renamed?  The name implies that it just checks whether the block is explicitly locked?  But in fact, this method updates the lock selection.

efischer19 replied:
I mean, it's a property setter, so when called as `modal.is_explicitly_locked = True`, it seems to indicate what's happening pretty clearly IMO. Do you have a suggestion for a better name?

nasthagiri replied:
Ah.. I missed that it's a property setter.  Yup - continue on.  Thanks.

nasthagiri commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -590,9 +578,11 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
-    StaffLockEditor = AbstractEditor.extend({
-        templateName: 'staff-lock-editor',
-        className: 'edit-staff-lock',
+    AbstractVisibilityEditor = AbstractEditor.extend({
```
Curious: What is the rationale for introducing an `AbstractVisibilityEditor`?

efischer19 replied:
All the logic in `AbstractVisibilityEditor` is shared by both `StaffLockEditor` (the `staff_only` checkbox with 2 states used by Section and Unit) and `ContentVisibilityEditor` (the new radio group with 3 states used by Subsection).

Just a bit of DRY-ing

nasthagiri commented:
Completed my pass.  Looks great overall.

catong commented on cms/templates/js/content-visibility-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+<form>
+<h3 class="modal-section-title" id="content_visibility_label"><%- gettext('Student Visibility') %></h3>
```
@efischer19 
This label "Student Visibility" reads as "visibility of students" to me - how about "Content Visibility" or "Subsection Visibility" if this label is always at the subsection level?

catong commented on cms/templates/js/content-visibility-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+        <li class="field-radio">
+            <label class="label">
+                <input class="input input-radio" name="content-visibility" type="radio" value="staff_only">
+                <%- gettext('Hide from students') %>
```
All labels should refer to "learners" rather than "students": 
"Hide from students" -> "Hide from learners"

catong replied:
Amending my own suggestion: we should also add a noun: "Hide content from learners" - for a more obvious comparison with the following option.

catong replied:
Should this option also have tip info? I think it would be helpful for course teams to understand the implications of hiding the content.

efischer19 replied:
I'll whip something up for a tip to see how that looks.

efischer19 replied:
Looking closer at your `Hide content from learners` suggestion - there's actually a large difference between this option and the one that follows.

- `Hide from learners` will hide the subsection's content *and* everything else about the subsection (from the student POV, it does not exist at all).
- `Hide content but not grades` (or however it ends up being labelled) will still allow students to see the subsection (it exists), but the contents are hidden via @nasthagiri's work on the LMS PR.

So including `content` was intended to make that differentiation clear.

What about options like:
- Hide Entire Subsection
- Hide Contents After Due Date
- Nothing Hidden

and then further details are given in the tip?

nasthagiri replied:
+1 on the wording suggestions.

catong commented on cms/templates/js/content-visibility-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+        <li class="field-radio">
+            <label class="label">
+                <input class="input input-radio" name="content-visibility" type="radio" value="hide_after_due">
+                <%- gettext('Hide content but not grades from students after due date') %>
```
Replace "students" with "learners"

catong replied:
I think the most important part of this option choice is "after due date", which gets a bit lost while people are parsing "hide content but not grades". What do you think about moving the "but not grades" into the tip info and just having the option read "Hide content from learners only after due date."
(The "don't affect grades while hiding" aspect is really our behavior fix and can be explained in the tip?)

efischer19 replied:
That's a good idea!

catong commented on cms/templates/js/content-visibility-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+        <li class="field-radio">
+            <label class="label">
+                <input class="input input-radio" name="content-visibility" type="radio" value="visible">
+                <%- gettext('Always show to students (Default)') %>
```
Replace "students" with "learners". 
Also, delete "(Default)" - not necessary.

catong commented on cms/templates/js/content-visibility-editor.underscore:
```diff
@@ -0,0 +1,33 @@
+
+    <% if (hasExplicitStaffLock && !ancestorLocked) { %>
+        <p class="tip tip-warning">
+            <% var message = gettext('If you make this %(xblockType)s visible to students, students will be able to see its content after the release date has passed and you have published the unit.'); %>
```
Replace "students" with "learners" ->
"... visible to learners, learners will be able to see ..."

catong replied:
Is this tip intended to be applicable for all 3 options? It appears only when I select "Hide content but not grades", and remains visible if I subsequently select "Always show to learners". It does not appear when I select "Hide from learners".

efischer19 replied:
This is a context sensitive tip.

If:
- the subsection is explicitly set to `staff_only` and
- there is no inherited lock from the containing section

Then:
- selecting an option that would make the content visible to students (either of the not `staff_only` choices)
makes the tip visible.

I think it's intended to be a warning for the situation of unhiding content. The same pattern is used here for the 2-state staff lock checkbox: https://github.com/edx/edx-platform/blob/efischer/hidden_studio/cms/templates/js/staff-lock-editor.underscore#L13-L28

catong replied:
Tip text also seems oddly large in comparison with the option labels themselves.

catong replied:
Thanks @efischer19. My brain found it difficult to change viewpoints from the "Hide" action offered by our options to the opposite "make visible" action referred to in the tip. Also, the "clear option" reference is confusing since we are not using checkboxes.
If I understand you correctly, we follow the pattern but don't use the same message text? If so, let's work out some alternative message text. I still need to think through the context-sensitive scenarios...

efischer19 replied:
@catong You're correct, but I don't believe there's any reason to go out of our way to use different message text. I'll make sure they match when I make my docstring update.

efischer19 replied:
Update: there are some small differences, due to subsections having more than one "visible to learners" option. I'll tag the diff when it gets pushed.

sstack22 commented:
@efischer19 - this looks amazing! Timed Exams are in enabled in your sandbox though, and they do not seem to be appearing: https://studio-efischer19.sandbox.edx.org/settings/advanced/course-v1:edX+DemoX+Demo_Course

Pre-requisites are working as expected though.

efischer19 commented:
@sstack22 My bad! There's a site-wide feature flag to enable timed exams that I haven't checked in, and I forgot to manually flip it on after this latest re-deploy. It's on now, and I can see timed exam options in studio.

efischer19 commented:
@catong Here's a screenshot of the new "tips" I added to the subsection menu - I just copied the style used by the timed exam editor for consistency. Does this styling seem better?

<img width="583" alt="screen shot 2016-07-19 at 4 32 24 pm" src="https://cloud.githubusercontent.com/assets/7373924/16965542/9b149a70-4dce-11e6-9dcf-3679fabc22a6.png">

efischer19 commented:
Also @nasthagiri @sanfordstudent - I've addressed all code feedback in a separate commit, ready for another round of review.

Keep in mind user-facing wording is still a WIP in many places.

catong commented:
@efischer19 This is awesome. The smaller font looks great, and thanks for putting together the basis for the new text!

Here are my suggestions for the options and help text. Per my conversation with @sstack22, I've also changed the order of the options so that the default comes first. 
@efischer19 @chris-mike @sstack22 @nasthagiri Your thoughts?

- Show entire subsection
Learners see the published subsection and can access its content. 
- Hide content after due date
After the subsection's due date has passed, learners can no longer access its content. The subsection remains included in grade calculations.
- Hide entire subsection
Learners do not see the subsection in the course outline. The subsection is not included in grade calculations.

efischer19 commented:
Updated screenshot, including a fix for the old warning looking terrible:

<img width="595" alt="screen shot 2016-07-19 at 5 23 24 pm" src="https://cloud.githubusercontent.com/assets/7373924/16967109/95ec136e-4dd5-11e6-825a-08467f112b56.png">

catong commented on cms/templates/js/content-visibility-editor.underscore:
```diff
@@ -0,0 +1,35 @@
+
+    <% if (hasExplicitStaffLock && !ancestorLocked) { %>
+        <p class="tip tip-warning">
+            <% var message = gettext('If you make this %(xblockType)s visible to learners, learners may be able to see its content after you have published the unit. Any units that are explicitly hidden from learners will remain hidden after you clear this option for the %(xblockType)s.'); %>
```
@efischer19 Any concerns with the variable XblockType vs hardcoded "subsection" in the option help text?
Can you clarify what is meant by "If you make this ```{subsection}``` visible to learners, learners may be able to see its content after you have published the unit." ?
Also thinking that "clear this option" should be something else since these are not checkboxes - and you can't just clear them all. Is it accurate to say, "Any units that are explicitly hidden from learners will remain hidden if you select "Show entire subsection." "?

efischer19 replied:
The gist of it is "If you change this value to anything other than `Hide entire subsection`, units that were previously hidden by inheriting that value may become visible once you publish"

I agree on `clear this option` being confusing, maybe `Any units that are explicitly hidden from learners will remain hidden regardless of the option you select for this subsection`?

catong replied:
@efischer19 For line 30 what do you think of this:
"Units inherit the visibility setting of the subsection they are in. If you make this subsection visible to learners, published units that were previously hidden are also visible. Only units that were explicitly hidden remain hidden regardless of the option you select for this subsection."

catong commented on cms/templates/course_outline.html:
```diff
@@ -199,7 +199,8 @@ <h3 class="title-3">${_("Setting release dates and grading policies")}</h3>
                 <p>${_("Select the Configure icon for a section or subsection to set its release date. When you configure a subsection, you can also set the grading policy and due date.")}</p>
                 <h3 class="title-3">${_("Changing the content students see")}</h3>
```
@efischer19  Would it be out of scope to change "students" to "learners" here? :-)

efischer19 replied:
Not at all! I'll get that fixed

catong commented on cms/templates/course_outline.html:
```diff
@@ -199,7 +199,8 @@ <h3 class="title-3">${_("Setting release dates and grading policies")}</h3>
                 <h3 class="title-3">${_("Changing the content students see")}</h3>
                 <p>${_("To publish draft content, select the Publish icon for a section, subsection, or unit.")}</p>
-                <p>${Text(_("To hide content from students, select the Configure icon for a section, subsection, or unit, then select {em_start}Hide from students{em_end}.")).format(em_start=HTML("<strong>"), em_end=HTML("</strong>"))}</p>
+                <p>${Text(_("To make a section, subsection, or unit unavailable to students (including grade calculations), select the Configure icon for a section, subsection, or unit, then select {em_start}Hide from students{em_end}.")).format(em_start=HTML("<strong>"), em_end=HTML("</strong>"))}</p>
```
@efischer19 Please bear with me - I'll work on this text.

efischer19 replied:
No worries - I'd rather be deliberate with these changes than rush them.

We should be able to have multiple links. I think that would provide a lot of value to course authors.

efischer19 replied:
@catong @sstack22 @chris-mike How does this look for multiple doc link options?

<img width="340" alt="screen shot 2016-07-20 at 12 20 05 pm" src="https://cloud.githubusercontent.com/assets/7373924/16994216/6ee8fd64-4e74-11e6-8400-d9d76b161a9b.png">

catong replied:
@efischer19 I love it!!

nasthagiri commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -605,43 +615,35 @@ define(["jquery", "edx-ui-toolkit/js/utils/spec-helpers/ajax-helpers", "common/j
-                selectAdvancedSettings = function() {
-                   this.$(".modal-section .settings-tab-button[data-tab='advanced']").click();
+                setContentVisibility = function (visibility) {
+                    if (visibility) {
```
Is this defensive line of code really needed?  Sometimes being overly defensive hides real issues.

efischer19 commented:
jenkins run lettuce

efischer19 commented:
@catong I had to make some changes to the context sensitive help code to get this working correctly - who would be a good reviewer for those changes? They're all contained in the latest commit.

Essentially, instead of defining a single link to be used everywhere on the page, I allow each "can I have a doc link please" call to specify a different key for the correct link.

catong commented:
@efischer19 I've worked with the Studio HTML pages and the help a little. I took a look and the mappings seem fine. The one thing I'm not clear about is the effect of "develop_course" in line 3, but we can test in your sandbox, I assume? Let me first help get the text finalized in that sidebar. Thanks!

catong commented:
Oh @efischer19 is the "return" value in line 3 the mapping for the Help link next to the username at the top right of the page? If so, what you have is a good choice. Thanks!

efischer19 commented on cms/templates/course_outline.html:
```diff
@@ -1,6 +1,6 @@
 <%page expression_filter="h"/>
 <%inherit file="base.html" />
-<%def name="online_help_token()"><% return "outline" %></%def>
+<%def name="online_help_token()"><% return "develop_course" %></%def>
```
@catong This help token isn't currently being used at all, I just wanted to provide a sane default in case some other link gets added to this page that wants to use the existing `${get_online_help_info(online_help_token())['doc_url']}` pattern. We can completely remove this line if you don't think that'll be an issue.

catong commented on cms/templates/course_outline.html:
```diff
@@ -193,16 +193,29 @@ <h2 class="sr">${_("Course Outline")}</h2>
+            <div class="bit">
+                <h3 class="title-3">${_("Changing the content learners see")}</h3>
+                <p>${_("To publish draft content, select the Publish icon for a section, subsection, or unit.")}</p>
+                <p>${Text(_("To make a section, subsection, or unit unavailable to learners (including grade calculations), select the Configure icon for a section, subsection, or unit, then select {em_start}Hide from learners{em_end}.")).format(em_start=HTML("<strong>"), em_end=HTML("</strong>"))}</p>
```
Suggested rewrite: "To make a section, subsection, or unit unavailable to learners, select the Configure icon for that level, then select the appropriate ```{em_start}```Hide```{em_end}``` option." Be aware that grades for hidden sections, subsections, and units are not included in grade calculations.

catong replied:
@efischer19 Here are my suggestions ^^ and below. I chose to generalize the reference to the Hide option, as the label is now different for subsections. Think it's OK?

catong commented on cms/templates/course_outline.html:
```diff
@@ -193,16 +193,29 @@ <h2 class="sr">${_("Course Outline")}</h2>
+                <h3 class="title-3">${_("Changing the content learners see")}</h3>
+                <p>${_("To publish draft content, select the Publish icon for a section, subsection, or unit.")}</p>
+                <p>${Text(_("To make a section, subsection, or unit unavailable to learners (including grade calculations), select the Configure icon for a section, subsection, or unit, then select {em_start}Hide from learners{em_end}.")).format(em_start=HTML("<strong>"), em_end=HTML("</strong>"))}</p>
+                <p>${Text(_("To hide content (but not grades) from learners after a due date, select the Configure icon for a subsection, then select {em_start}Hide contents but not grades from learner after due date{em_end}.")).format(em_start=HTML("<strong>"), em_end=HTML("</strong>"))}</p>
```
To hide the content of a subsection from learners after the subsection due date has passed, select the Configure icon for a subsection, then select ```{em_start}```Hide Content After Due Date```{em_end}```. Grades for the subsection remain included in grade calculations.

nasthagiri commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -801,7 +790,7 @@ define(["jquery", "edx-ui-toolkit/js/utils/spec-helpers/ajax-helpers", "common/j
                     ]);
                     createCourseOutlinePage(this, mockCourseJSON, false);
-                    outlinePage.$('.outline-subsection .configure-button').click();
+                    outlinePage.$('.outline-unit .configure-button').click();
```
Why did this need to change to unit from subsection?

efischer19 replied:
This is testing `does not show settings tab headers if there is only one tab to show`. As a result of the tab movement, subsections now always have 2 tabs (`Basic` and `Advanced`), so this test needs to be run on a unit, which only has the one tab (there's nothing on its `Basic` tab, so the visibility-editor-on-advanced gets collapsed down to the single untabbed view)

catong commented on cms/templates/js/staff-lock-editor.underscore:
```diff
@@ -7,27 +7,18 @@
+                    <%= gettext('If the unit was previously published and released to learners, any changes you made to the unit when it was hidden will now be visible to learners.') %>
                   <% } else { %>
-                    <% var message = gettext('If you make this %(xblockType)s visible to students, students will be able to see its content after the release date has passed and you have published the unit.'); %>
+                    <% var message = gettext('If you make this %(xblockType)s visible to learners, learners will be able to see its content after the release date has passed and you have published the unit. Any units that are explicitly hidden from learners will remain hidden after you clear this option for the %(xblockType)s.') %>
```
In the last sentence, perhaps "Only units..." instead of "Any units..."?

nasthagiri commented:
👍  Looks great!  A lot of additional awesome changes coming from this PR!

efischer19 commented:
jenkins run a11y

catong commented:
@efischer19 This is really awesome. 👍 
I will open a separate ticket to request to update the modal headings at the section and unit levels - they both currently read "Student Visibility", which we changed in this PR.

catong commented on cms/templates/js/publish-xblock.underscore:
```diff
@@ -97,10 +97,14 @@ var visibleToStaffOnly = visibilityState === 'staff_only';
         </ul>
+        <p>
+            <%- gettext("Note: Do not hide graded assignments after they have been released.") %>
+            <a href="#"><%- gettext("Link to documentation.") %></a>
```
@efischer19 I'm now thinking this link to doc isn't necessary. Can we remove? Thanks!

efischer19 commented on cms/templates/js/staff-lock-editor.underscore:
```diff
@@ -1,31 +1,28 @@
+    <% } else { %>
+        <%- gettext('Section Visibility') %>
+    <% } %>
+</h3>
```
@catong 😉

catong replied:
👯

efischer19 commented:
Final commit was to address Carol's last comment, I've now squash/rebased and will merge once tests have passed.

efischer19 commented:
jenkins run lettuce

efischer19 commented:
jenkins run lettuce

sstack22 commented:
@efischer19 , @catong  - I just noticed that the text when you change from "Hide entire Subsection" only makes sense if you are changing it to "Show Entire Subsection".  So, meaning, when I toggle from "Hide Entire Subsection" to "Hide Content After Due Date" I get the below text, which is confusing:

"Units inherit the visibility setting of the subsection they are in. If you make this subsection visible to learners, published units that were previously hidden also become visible. Only units that were explicitly hidden remain hidden regardless of the option you select for this subsection."

catong commented:
@efischer19 @sstack22 
"Hide Content After Due Date" does make the subsection visible, except after the due date, so the message is true, but yes, could be confusing. 
It can be difficult to decide on the right amount of guidance in the UI. Shelby, once course teams understand each setting, esp. with updated documentation, do you think course teams will still be confused by this message?

sstack22 commented:
@catong @efischer19  - I think it would make the most sense to not display any message when toggling from hidden, to hide content after due date.

efischer19 commented:
@sstack22 that also raises the question - should toggling from `hide_after_due` to `visible` display the message?

I'll get a new PR filed so we can address this before a new release is cut tomorrow morning

cahrens commented on cms/templates/js/content-visibility-editor.underscore:
```diff
@@ -0,0 +1,34 @@
+    <ul class="list-fields list-input content-visibility" role="group" aria-labelledby="content_visibility_label">
+        <li class="field-radio">
+            <label class="label">
+                <input class="input input-radio" name="content-visibility" type="radio" value="visible" aria-described-by"visible_description">
```
"aria-described-by" is not a valid ARIA attribute. It is "aria-describedby". You can find this sort of error using Chrome's built-in accessibility Audit (in Chrome developer tools).

Also there is a typo in line 7 in that there is no equal sign.

I also don't understand the usage of ul/li with radio buttons inside of li's. I'll open a JIRA ticket for the issues. I imagine this HTML was likely copied from somewhere else, just as another PR has now copied this!

cahrens commented on cms/templates/js/timed-examination-preference-editor.underscore:
```diff
@@ -52,14 +52,6 @@
                 </label>
                 <p class='field-message' id='review-rules-description'><%- gettext('Specify any additional rules or rule exceptions that the proctoring review team should enforce when reviewing the videos. For example, you could specify that calculators are allowed.') %></p>
             </li>
-            <li class="field-checkbox field-hide-after-due">
```
@catong It appears that this documentation did not get updated for the removal of this option as a Timed Exam-specific option.

http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/course_features/timed_exams.html

catong replied:
Thanks, @cahrens I'll take a look.
